lotus-soup/paych/stress.go

```diff
@@ -59,102 +74,236 @@ func Stress(t *testkit.TestEnvironment) error {
 	cancel()
 
 	switch mode {
-	case "receiver":
-		// one receiver, everyone else is a sender.
-		<-t.SyncClient.MustBarrier(ctx, SendersDoneState, t.TestGroupInstanceCount-1).C
-
-	case "sender":
-		// we're going to lock up all our funds into this one payment channel.
-		recv := clients[0]
-		balance, err := cl.FullApi.WalletBalance(ctx, cl.Wallet.Address)
+	case ModeReceiver:
+		err := runReceiver(t, ctx, cl)
+		if err != nil {
+			return err
+		}
+
+	case ModeSender:
+		err := runSender(ctx, t, clients, cl)
 		if err != nil {
-			return fmt.Errorf("failed to acquire wallet balance: %w", err)
+			return err
 		}
+	}
+
+	t.SyncClient.MustSignalEntry(ctx, testkit.StateDone)
```

Comment:
@raulk for some reason the bootstrapper never exits, even though I'm signalling StateDone here

Reply:
Right! I think this is because the miners have not signalled `StateDone` (the boostrapper waits for all instances to signal).

The shutdown procedure is:

- clients are done, and signal on `StateDone`, and also on `StateStopMining` to inform miners that they can stop mining now.
- miners wait for all clients to signal on `StateStopMining`, and only then do they signal on `StateDone`.
- by now, all nodes should've signalled on `StateDone`, and the bootstrapper will shut down.